### ISSUE TITLE
Problems with bytea performance

### ISSUE THREAD

**Issue by mramato** [tgriesser/knex#2052]
I have the below simple query:

```javascript
return client('models')
    .first('glb')
    .where({
        assets_id: assetId
    })
    .then(function (row) {
        if (defined(row)) {
            return row.glb;
        }
    });
```

glb is a `binary` type (`bytea` since I'm using postgres 9.5) and contains a single row where the `glb` column has an entry that is `28630077` in bytes (~27MB).  The query takes `13305 ms` to run (time until `then` is called) and the Node process (not the DB process) maxes out the CPU while the query is running). If I query for the `assets_id` column instead of the `glb` column, it only takes `2 ms`.

Running the same query with the same data from the psql command line completes almost immediately:

```bash
time psql -A -c "select glb from models where assets_id=16" master3 postgres > out.glb

real    0m0.679s
user    0m0.000s
sys     0m0.031s
```

Here's the table setup for completeness

```javascript
return knex.schema.createTable('models', function (table) {
    table.integer('assets_id').primary().references('assets.id').onDelete('CASCADE');
    table.binary('glb').notNullable();
});
```

Am I doing something wrong?  Or is there a performance issue with knex and `bytea` column types?  I'd be happy to debug this myself if someone can point me in the correct direction.  This is using the latest released version of knex.  Thanks.

**Comment by elhigu** [comment]
Could you please check out if the query is also really slow when using `node-postgres` driver directly?

**Comment by mramato** [comment]
Thanks.  I can confirm that using `node-postgres` directly has the same issue and running time.

I decided to also  check `pg-native`  while I was at it and it's quite a bit faster (~440ms), but from what I can tell, it's no longer possible to use pg-native with knex (and I'm not sure if that's a viable strategy to begin with).

Does this mean I'm out of luck? Any guidance you can provide would be greatly appreciated.

**Comment by elhigu** [comment]
Thanks for checking. 

I think that better place to discuss it would be `node-postgres` issue tracker. This sounds like one of the rare cases where using `pg-native` is actually reasonable (if `node-postgres` cannot be fixed to work faster).

Support for `pg-native` was dropped, because it didn't support all API's that knex uses... maybe that decision could be re-evaluated...

**Comment by mramato** [comment]
Thanks, I opened https://github.com/brianc/node-postgres/issues/1286 to hopefully get some guidance from them.  Feel free to close this unless you want to use it as a discussion point for re-adding `pg-native` support.

Assuming `pg-native` ends up being my only option, I would be happy to help re-add native support, otherwise I'm not sure what my options are.

**Comment by mramato** [comment]
@elhigu, great news.  I was able to track down the problem and open a PR: https://github.com/brianc/node-packet-reader/pull/3  Assuming that gets merged and published, it looks like it will be a significant improvement for `node-postgres` read performance all-around.

**Comment by elhigu** [comment]
@mramato I just read the `node-postgres` issue through, thats just plain awesomeness that you were able to track it down! Hope that it ends up to upstream soon. I'm closing this now, but let me know if there is anything more I can do for this.

**Comment by mramato** [comment]
Thanks, and thanks for all of the help.

**Comment by hansef** [comment]
This has been merged:

* https://github.com/brianc/node-postgres/issues/1286
* https://github.com/brianc/node-packet-reader/pull/3

It's now fixed in node-pg 6.2.2 - any chance we can get a version bump on the dependency in knex soon?

**Comment by mramato** [comment]
@hansef there's no need for a `knex` version bump because `pg` is an assumed sibling dependency, not a direct one.  Just make sure you depend on the latest `pg` in your own application and you will pick up the change.  `knex` only has a dev dependency on "pg" for it's own unit tests and it's not part of the production app.
